Eliminate `word_and_empty` methods.

nnethercote opened the pull request:
To remove the last remaining `Ident::empty` uses.

r? @jdonszelmann

nnethercote committed:
Eliminate `word_or_empty` methods.

To get rid of the `Ident::empty` uses.

This requires introducing `PathParser::word_sym`, as an alternative to
`PathParser::word`.

nnethercote commented:
An opinionated PR, this one... sorry there are a bunch of changes in a single commit, it was hard to separate them out.

jdonszelmann commented:
I see, well I'll take a look :)

nnethercote commented:
This helps with #137978.

jdonszelmann commented:
hm, would you mind at least making the removals from parser.rs a separate commit. I'm not 100% sure I agree with those removals. I intended those APIs to be for all the future attribute parsers that will be converted to a dedicated parser implementation. The reason I am kind of okay with it is because that some of those APIs were made before I allowed easy matching on attribute parsers, which really is a better API so maybe that's what we should switch to. But that'd still be nicer to discuss/refer to if it were a separate commit. I also quickly want to see whether some of my open but blocked PRs relied on those APIs just to see if I had very good reasons for them but I doubt it.

jdonszelmann commented on compiler/rustc_attr_parsing/src/context.rs:
```diff
@@ -260,7 +260,8 @@ impl<'sess> AttributeParser<'sess> {
                 // }
                 ast::AttrKind::Normal(n) => {
                     let parser = MetaItemParser::from_attr(n, self.dcx());
-                    let (path, args) = parser.deconstruct();
```
I'm not sure this is an improvement

jdonszelmann replied:
since it's a reasonably common operation I think

nnethercote replied:
It doesn't have many uses, which is why the second commit got rid of it. But no matter, I've reinstated the old code here, we can discuss `deconstruct` more in the follow-up.

jdonszelmann commented on compiler/rustc_attr_parsing/src/attributes/repr.rs:
```diff
@@ -96,57 +96,65 @@ fn parse_repr(cx: &AcceptContext<'_>, param: &MetaItemParser<'_>) -> Option<Repr
+        ) => {
             cx.emit_err(session_diagnostics::InvalidReprHintNoParen {
                 span: param.span(),
-                name: ident.to_string(),
```
well, in some cases I guess it's a little annoying since we match on multiple symbols, and know for sure its one of those so the unwrap can't fail...

nnethercote replied:
The unwraps can be avoided with `@` patterns. I'll do that.

jdonszelmann commented on compiler/rustc_attr_parsing/src/attributes/deprecation.rs:
```diff
@@ -13,16 +12,13 @@ pub(crate) struct DeprecationParser;
 fn get(
     cx: &AcceptContext<'_>,
-    ident: Ident,
+    name: Symbol,
```
it surprised me we never use the span of this ident. But I guess we usually use the param_span. Well, nice!

jdonszelmann commented:
The first commit is blocking another PR. That's the one I reviewed and am happy with merging (maybe after some nitpicks are addressed). I think some of those unwraps might be unnecessary. I can take one more look tomorrow if you decide to change something and then we can merge it. The other, second, commit I have some doubts about right now, which I expressed earlier. If it helps unblock another PR, feel free to refile the second commit in another PR and assign me, then I'll take a look later when I have more time. Rust Week planning is taking a lot. Good night!

nnethercote committed:
Avoid some unwraps.

By using `@` patterns more.

Also, use `Symbol` more in a couple of errors to avoid some unnecessary
conversions to strings. This even removes a lifetime.

nnethercote commented:
I have removed the original second commit with the API changes; we can try again with those in a follow-up PR.

I have added a new second commit that avoids the unwraps and does a couple of other small related improvements.

@rustbot ready

jdonszelmann commented:
Nice, I like it :) And yea, I think the other changes are generally good, just can't review them properly atm because of rust week. We'll get back to it!


@bors r+ rollup

bors commented:
:pushpin: Commit 603766cd727a5c8b8151a542ab53cf8f84a38f4f has been approved by `jdonszelmann`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jdonszelmann 603766cd727a5c8b8151a542ab53cf8f84a38f4f -->
<!-- homu: {"type":"Approved","sha":"603766cd727a5c8b8151a542ab53cf8f84a38f4f","approver":"jdonszelmann","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
